src/dali/plugin/input/rest/kucoin_futures.py

```diff
@@ -0,0 +1,288 @@
+
+    def _initialize_client(self) -> kucoinfutures:
+        return kucoinfutures(
+            {
+                "apiKey": self.__api_key,
+                "secret": self.__api_secret,
+                "password": self.__api_passphrase,
+                "enableRateLimit": True,
+            }
+        )
+
+    def exchange_name(self) -> str:
+        return self.__EXCHANGE_NAME
+
+    def plugin_name(self) -> str:
+        return self.__PLUGIN_NAME
+
+    @property
+    def _client(self) -> kucoinfutures:
+        super_client: Exchange = super()._client
+        if not isinstance(super_client, kucoinfutures):
+            raise TypeError("super_client is not of type kucoinfutures")
+
+        return super_client
+
+    # using kucoin transaction history api instead. (implicit api)
+    def _get_process_deposits_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        return None
+
+    def _get_process_withdrawals_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        return None
+
+    def _get_process_trades_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        return None
+
+    def _process_gains(
+        self,
+        in_transactions: List[InTransaction],
+        out_transactions: List[OutTransaction],
+    ) -> None:
+        pass
+
+    def _fetch_data(
+        self,
+        start_at: int,
+        end_at: int,
+    ) -> List[Dict[str, Any]]:
+
+        offset: int = 1
+        retries: int = 0
+        has_more: bool = False
+
+        ledger_transactions: Dict[str, Any] = {}
+        items: List[Dict[str, Any]] = []
+
+        while (start_at <= end_at) and retries < 4:
+            try:
+                ledger_transactions = self._client.futuresPrivateGetTransactionHistory({_START_AT: start_at, _END_AT: start_at + _TS_INCREMENT})
+                data_list = ledger_transactions[_DATA][_DATA_LIST]
+                has_more = ledger_transactions[_DATA][_HAS_MORE]
+
+                for item in data_list:
+                    items.append(item)
+                    offset = item[_OFFSET]
+
+                # https://github.com/ccxt/ccxt/issues/10273
+                # enableRateLimit is not working for kucoin
+                time.sleep(0.15)
+                retries = 0
+
+                while has_more and (retries < 4):
+                    try:
+                        ledger_transactions = self._client.futuresPrivateGetTransactionHistory(
+                            {_START_AT: start_at, _END_AT: start_at + _TS_INCREMENT, _OFFSET: offset}
+                        )
+                        data_list = ledger_transactions[_DATA][_DATA_LIST]
+                        has_more = ledger_transactions[_DATA][_HAS_MORE]
+
+                        for item in data_list:
+                            items.append(item)
+                            offset = item[_OFFSET]
+
+                        time.sleep(0.15)
+
+                    except RateLimitExceeded:
```

Comment:
To make this more robust, you could add exceptions for DDoS and network errors. There is an example of this in the [CCXT pricing plugin](https://github.com/eprbell/dali-rp2/blob/cc2bb28c9d055da7e95a7fc2c82dd0646bd4a70b/src/dali/plugin/pair_converter/ccxt.py#L334-L350).

Reply:
fixed